File: weather-batch-scala/src/main/scala/com/myorg/WeatherBatchService.scala
package com.myorg

import io.circe._
import sttp.client3._
import sttp.client3.httpclient.HttpClientSyncBackend
import sttp.model.MediaType

import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale
import scala.util.Try

case class Forecast(
  publishingOffice: String,
  reportDatetime: OffsetDateTime,
  todayForecast: DailyForecast,
  tomorrowForecast: DailyForecast,
)

case class DailyForecast(timeDefine: OffsetDateTime, content: String)

object WeatherBatchService {
  private val WEATHER_URL          = "https://www.jma.go.jp/bosai/forecast/data/forecast"
  private val YOKOHAMA_OFFICE_CODE = "140000"
  private val TYPETALK_TOPIC_URL   = "https://typetalk.com/api/v1/topics/259767"

  private val dateTimeFormat = DateTimeFormatter.ofPattern("yyyy年MM月dd日 HH時", Locale.JAPANESE)
  private val dateFormat     = DateTimeFormatter.ofPattern("dd日 (EEE) ", Locale.JAPANESE)
}

class WeatherBatchService(typetalkSettings: TypetalkSettings) {

  import WeatherBatchService._

  private[this] val backend = HttpClientSyncBackend()

  def run(): Either[Throwable, Unit] = {
    for {
      weatherJson <- fetchYokohamaWeather
      forecast    <- extractYokohamaData(weatherJson)
      _           <- sendToTypetalk(forecast)
    } yield ()
  }

  private def fetchYokohamaWeather: Either[Throwable, Json] = {
    val request = basicRequest
      .get(uri"${WEATHER_URL}/${YOKOHAMA_OFFICE_CODE}.json")

    val response = request.send(backend)

    response.body.fold(
      err => Left(new IllegalStateException(err)),
      parser.parse,
    )
  }

  private def extractYokohamaData(weatherJson: Json): Either[Throwable, Forecast] = {
    val c           = weatherJson.hcursor
    val details     = c.downN(0)
    val timeSeries  = details.downField("timeSeries").downN(0)
    val timeDefines = timeSeries.downField("timeDefines")
    val weathers    = timeSeries.downField("areas").downN(0).downField("weathers")
    val getForecast = (i: Int) => {
      for {
        dateString <- timeDefines.downN(i).as[String]
        date       <- Try(OffsetDateTime.parse(dateString)).toEither
        content    <- weathers.downN(i).as[String]
      } yield DailyForecast(date, content)
    }

    for {
      publishingOffice     <- details.get[String]("publishingOffice")
      reportDatetimeString <- details.get[String]("reportDatetime")
      reportDatetime       <- Try(OffsetDateTime.parse(reportDatetimeString)).toEither
      todayForecast        <- getForecast(0)
      tomorrowForecast     <- getForecast(1)
    } yield Forecast(publishingOffice, reportDatetime, todayForecast, tomorrowForecast)
  }

  private def sendToTypetalk(forecast: Forecast): Either[Throwable, String] = {
    val todayForecast    = forecast.todayForecast
    val tomorrowForecast = forecast.tomorrowForecast

    val message =
      s"""横浜の天気
         |${dateTimeFormat.format(forecast.reportDatetime)} ${forecast.publishingOffice} 発表 (気象庁より)
         |今日 ${dateFormat.format(todayForecast.timeDefine)} ：  ${todayForecast.content}
         |明日 ${dateFormat.format(tomorrowForecast.timeDefine)} ：  ${tomorrowForecast.content}
         |""".stripMargin
    val json = Json.obj("message" -> Json.fromString(message)).noSpaces
    val request = basicRequest
      .post(uri"$TYPETALK_TOPIC_URL")
      .header("X-TYPETALK-TOKEN", typetalkSettings.token)
      .contentType(MediaType.ApplicationJson)
      .body(json)
    val response = request.send(backend)

    response.body.fold(
      err => Left(new IllegalStateException(err)),
      Right(_),
    )
  }
}

File: weather-batch-java/src/main/java/com/myorg/weather/batch/WeatherBatchService.java
package com.myorg.weather.batch;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import javax.enterprise.context.ApplicationScoped;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;
import java.util.Map;

@ApplicationScoped
public class WeatherBatchService {

    public record Forecast(
      String publishingOffice,
      OffsetDateTime reportDatetime,
      DailyForecast todayForecast,
      DailyForecast tomorrowForecast,
      DailyForecast dayAfterTomorrowForecast
    ) {
    }

    public record DailyForecast(OffsetDateTime timeDefine, String content) {
    }

    private static final String WEATHER_URL = "https://www.jma.go.jp/bosai/forecast/data/forecast";
    private static final String YOKOHAMA_OFFICE_CODE = "140000";
    private static final String TYPETALK_TOPIC_URL = "https://typetalk.com/api/v1/topics/259767";

    private static final ObjectMapper mapper = new ObjectMapper();

    private static final DateTimeFormatter dateTimeFormat = DateTimeFormatter.ofPattern("yyyy年MM月dd日 HH時", Locale.JAPANESE);
    private static final DateTimeFormatter dateFormat = DateTimeFormatter.ofPattern("dd日 (EEE) ", Locale.JAPANESE);

    private final HttpClient client = HttpClient.newHttpClient();

    private final String typetalkToken;

    public WeatherBatchService(@ConfigProperty(name = "typetalk.token") String typetalkToken) {
        this.typetalkToken = typetalkToken;
    }

    public void run() throws IOException, InterruptedException {
        var weatherJson = fetchYokohamaWeather();
        var aaa = extractYokohamaData(weatherJson);
        sendToTypetalk(aaa);
    }

    private JsonNode fetchYokohamaWeather() throws IOException, InterruptedException {
        var request = HttpRequest.newBuilder()
          .uri(URI.create(WEATHER_URL + "/" + YOKOHAMA_OFFICE_CODE + ".json"))
          .build();

        var response = this.client.send(request, HttpResponse.BodyHandlers.ofInputStream());
        return mapper.readTree(response.body());
    }

    private Forecast extractYokohamaData(JsonNode weatherJson) {
        var details = weatherJson.get(0);
        var publishingOffice = details.get("publishingOffice").asText();
        var reportDatetime = OffsetDateTime.parse(details.get("reportDatetime").asText());
        var timeSeries = details.get("timeSeries").get(0);
        var timeDefines = timeSeries.get("timeDefines");
        var weathers = timeSeries.get("areas").get(0).get("weathers");

        var todayForecast = new DailyForecast(
          OffsetDateTime.parse(timeDefines.get(0).asText()),
          weathers.get(0).asText()
        );
        var tomorrowForecast = new DailyForecast(
          OffsetDateTime.parse(timeDefines.get(1).asText()),
          weathers.get(1).asText()
        );
        var dayAfterTomorrowForecast = new DailyForecast(
          OffsetDateTime.parse(timeDefines.get(2).asText()),
          weathers.get(2).asText()
        );
        return new Forecast(publishingOffice, reportDatetime, todayForecast, tomorrowForecast, dayAfterTomorrowForecast);
    }

    private void sendToTypetalk(Forecast forecast) throws IOException, InterruptedException {
        var todayForecast = forecast.todayForecast();
        var tomorrowForecast = forecast.tomorrowForecast();
        var dayAfterTomorrowForecast = forecast.dayAfterTomorrowForecast();
        var message = """
          横浜の天気
          %s %s 発表 (気象庁より)
          今日　 %s ：  %s
          明日　 %s ：  %s
          明後日 %s ：  %s
          """
          .formatted(
            dateTimeFormat.format(forecast.reportDatetime()), forecast.publishingOffice(),
            dateFormat.format(todayForecast.timeDefine()), todayForecast.content(),
            dateFormat.format(tomorrowForecast.timeDefine()), tomorrowForecast.content(),
            dateFormat.format(dayAfterTomorrowForecast.timeDefine()), dayAfterTomorrowForecast.content()
          );

        var json = mapper.writeValueAsString(Map.of("message", message));
        var request = HttpRequest.newBuilder()
          .uri(URI.create(TYPETALK_TOPIC_URL))
          .header("Content-Type", "application/json")
          .header("X-TYPETALK-TOKEN", this.typetalkToken)
          .POST(HttpRequest.BodyPublishers.ofString(json))
          .build();

        this.client.send(request, HttpResponse.BodyHandlers.ofString());
    }
}
